examples/lpt_nbody_demo.py

```diff
@@ -155,61 +133,57 @@ def _sample_gaussian():
 def cic_paint(displacement, halo_size):
   """ Paints particles on a mesh using Cloud-In-Cell interpolation.
 
-    Parameters
-    ----------
-    displacement : jnp.ndarray
-        Displacement field of particles.
-    sharding : Any
-        Sharding strategy for the array.
-    halo_size : int
-        Halo size for painting.
-
-    Returns
-    -------
-    jnp.ndarray
-        Density field.
+    Args:
+        displacement (jnp.ndarray): Displacement of each particle.
+        halo_size (int): Halo size for painting.
+
+    Returns:
+        jnp.ndarray: Density field.
   """
-  local_mesh_shape = _global_to_local_size(displacement.shape)
+  local_mesh_shape = _global_to_local_size(displacement.shape[0])
+  hs = halo_size
 
-  @partial(shmap, in_specs=(P('z', 'y'),), out_specs=P('z', 'y'))
+  @partial(shmap, in_specs=(P('x', 'y'),), out_specs=P('x', 'y'))
   def cic_op(disp):
     """ CiC operation on each local slice of the mesh."""
     # Create a mesh to paint the particles on for the local slice
     mesh = jnp.zeros(disp.shape[:-1], dtype='float32')
 
-    # Padding the output array along the two first dimensions
-    mesh = jnp.pad(mesh,
-                   [[halo_size, halo_size], [halo_size, halo_size], [0, 0]])
+    # Padding the mesh along the two first dimensions
+    mesh = jnp.pad(mesh, [[hs, hs], [hs, hs], [0, 0]])
 
-    a, b, c = jnp.meshgrid(
+    # Compute the position of the particles on a regular grid
+    pos_x, pos_y, pos_z = jnp.meshgrid(
         jnp.arange(local_mesh_shape[0]),
         jnp.arange(local_mesh_shape[1]),
         jnp.arange(local_mesh_shape[2]),
         indexing='ij')
 
     # adding an offset of size halo size
-    pmid = jnp.stack([a + halo_size, b + halo_size, c], axis=-1)
-    return scatter(pmid.reshape([-1, 3]), disp.reshape([-1, 3]), mesh)
+    pos = jnp.stack([pos_x + hs, pos_y + hs, pos_z], axis=-1)
+
+    # Apply scatter operation to paint the particles on the local mesh
+    field = scatter(pos.reshape([-1, 3]), disp.reshape([-1, 3]), mesh)
 
-  # Performs painting on padded mesh
+    return field
+
+  # Performs painting on a padded mesh, with halos on the two first dimensions
   field = cic_op(displacement)
 
   # Run halo exchange to get the correct values at the boundaries
   field = jaxdecomp.halo_exchange(
       field,
-      halo_extents=(halo_size // 2, halo_size // 2, 0),
+      halo_extents=(hs // 2, hs // 2, 0),
       halo_periods=(True, True, True))
 
-  @partial(shmap, in_specs=(P('z', 'y'),), out_specs=P('z', 'y'))
+  @partial(shmap, in_specs=(P('x', 'y'),), out_specs=P('x', 'y'))
```

Comment:
The axes were called Z and Y according to cudecomp
The fastest axis is called X and it is layed out like this ZYX (X pencil)
It is just a naming convention so not important

Reply:
Yes, but here I'm thinking about user expectations. A naive user would want to call the dimensions "x", "y", "z" in that order. 

In general, it's always better to design the code such that it aligns with what a user that has not read the documentation would expect and avoid user surprise.

So, even if this goes against the nomenclature in cuDecomp, I do think it's a better choice for the end-user.